File: app/src/main/java/com/babestudios/companyinfouk/data/network/CompaniesHouseService.java
package com.babestudios.companyinfouk.data.network;

import com.babestudios.companyinfouk.BuildConfig;
import com.babestudios.companyinfouk.data.model.charges.Charges;
import com.babestudios.companyinfouk.data.model.company.Company;
import com.babestudios.companyinfouk.data.model.filinghistory.FilingHistoryList;
import com.babestudios.companyinfouk.data.model.insolvency.Insolvency;
import com.babestudios.companyinfouk.data.model.officers.Officers;
import com.babestudios.companyinfouk.data.model.officers.appointments.Appointments;
import com.babestudios.companyinfouk.data.model.persons.Persons;
import com.babestudios.companyinfouk.data.model.search.CompanySearchResult;

import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;

public interface CompaniesHouseService {
	@GET(BuildConfig.COMPANIES_HOUSE_SEARCH_COMPANIES_ENDPOINT)
	Observable<CompanySearchResult> searchCompanies(@Header("Authorization") String authorization,
													@Query("q") String searchTerm,
													@Query("items_per_page") String itemsPerPage,
													@Query("start_index") String startIndex);

	@GET(BuildConfig.COMPANIES_HOUSE_GET_COMPANY_ENDPOINT)
	Observable<Company> getCompany(@Header("Authorization") String authorization,
								   @Path("companyNumber") String companyNumber);


	@GET(BuildConfig.COMPANIES_HOUSE_GET_FILING_HISTORY_ENDPOINT)
	Observable<FilingHistoryList> getFilingHistory(@Header("Authorization") String authorization,
												   @Path("companyNumber") String companyNumber,
												   @Query("category") String category,
												   @Query("items_per_page") String itemsPerPage,
												   @Query("start_index") String startIndex);

	@GET(BuildConfig.COMPANIES_HOUSE_GET_CHARGES_ENDPOINT)
	Observable<Charges> getCharges(@Header("Authorization") String authorization,
								   @Path("companyNumber") String companyNumber,
								   @Query("items_per_page") String itemsPerPage,
								   @Query("start_index") String startIndex);

	@GET(BuildConfig.COMPANIES_HOUSE_GET_INSOLVENCY_ENDPOINT)
	Observable<Insolvency> getInsolvency(@Header("Authorization") String authorization,
										 @Path("companyNumber") String companyNumber);

	@GET(BuildConfig.COMPANIES_HOUSE_GET_OFFICERS_ENDPOINT)
	Observable<Officers> getOfficers(@Header("Authorization") String authorization,
									 @Path("companyNumber") String companyNumber,
									 @Query("registerView") String registerView,
									 @Query("registerType") String registerType,
									 @Query("orderBy") String orderBy,
									 @Query("items_per_page") String itemsPerPage,
									 @Query("start_index") String startIndex);

	@GET(BuildConfig.COMPANIES_HOUSE_GET_OFFICERS_APPOINTMENTS_ENDPOINT)
	Observable<Appointments> getOfficerAppointments(@Header("Authorization") String authorization,
													@Path("officerId") String officerId,
													@Query("items_per_page") String itemsPerPage,
													@Query("start_index") String startIndex);

	@GET(BuildConfig.COMPANIES_HOUSE_GET_PERSONS_ENDPOINT)
	Observable<Persons> getPersons(@Header("Authorization") String authorization,
								   @Path("companyNumber") String companyNumber,
								   @Query("registerView") String registerView,
								   @Query("items_per_page") String itemsPerPage,
								   @Query("start_index") String startIndex);
}

File: data/src/main/java/com/babestudios/companyinfouk/data/network/CompaniesHouseService.kt
package com.babestudios.companyinfouk.data.network

import com.babestudios.companyinfouk.data.BuildConfig
import com.babestudios.companyinfouk.data.model.charges.ChargesDto
import com.babestudios.companyinfouk.data.model.company.CompanyDto
import com.babestudios.companyinfouk.data.model.filinghistory.FilingHistoryDto
import com.babestudios.companyinfouk.data.model.insolvency.InsolvencyDto
import com.babestudios.companyinfouk.data.model.officers.AppointmentsResponseDto
import com.babestudios.companyinfouk.data.model.officers.OfficersResponseDto
import com.babestudios.companyinfouk.data.model.persons.PersonDto
import com.babestudios.companyinfouk.data.model.persons.PersonsResponseDto
import com.babestudios.companyinfouk.domain.model.search.CompanySearchResult
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

internal interface CompaniesHouseService {
	@GET(BuildConfig.COMPANIES_HOUSE_SEARCH_COMPANIES_ENDPOINT)
	suspend fun searchCompanies(
		@Query("q") searchTerm: String,
		@Query("items_per_page") itemsPerPage: String,
		@Query("start_index") startIndex: String,
	): CompanySearchResult

	@GET(BuildConfig.COMPANIES_HOUSE_GET_COMPANY_ENDPOINT)
	suspend fun getCompany(
		@Path("companyNumber") companyNumber: String,
	): CompanyDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_FILING_HISTORY_ENDPOINT)
	suspend fun getFilingHistory(
		@Path("companyNumber") companyNumber: String,
		@Query("category") category: String,
		@Query("items_per_page") itemsPerPage: String,
		@Query("start_index") startIndex: String,
	): FilingHistoryDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_CHARGES_ENDPOINT)
	suspend fun getCharges(
		@Path("companyNumber") companyNumber: String,
		@Query("items_per_page") itemsPerPage: String,
		@Query("start_index") startIndex: String,
	): ChargesDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_INSOLVENCY_ENDPOINT)
	suspend fun getInsolvency(
		@Path("companyNumber") companyNumber: String,
	): InsolvencyDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_OFFICERS_ENDPOINT)
	@Suppress("LongParameterList")
	suspend fun getOfficers(
		@Path("companyNumber") companyNumber: String,
		@Query("registerView") registerView: String?,
		@Query("registerType") registerType: String?,
		@Query("orderBy") orderBy: String?,
		@Query("items_per_page") itemsPerPage: String,
		@Query("start_index") startIndex: String,
	): OfficersResponseDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_OFFICERS_APPOINTMENTS_ENDPOINT)
	suspend fun getOfficerAppointments(
		@Path("officerId") officerId: String,
		@Query("items_per_page") itemsPerPage: String,
		@Query("start_index") startIndex: String,
	): AppointmentsResponseDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_PERSONS_ENDPOINT)
	suspend fun getPersons(
		@Path("companyNumber") companyNumber: String,
		@Query("registerView") registerView: String?,
		@Query("items_per_page") itemsPerPage: String,
		@Query("start_index") startIndex: String,
	): PersonsResponseDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_PERSONS_INDIVIDUAL_ENDPOINT)
	suspend fun getPersonIndividual(
		@Path("companyNumber") companyNumber: String,
		@Path("pscId") pscId: String,
	): PersonDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_PERSONS_CORPORATE_ENDPOINT)
	suspend fun getCorporatePerson(
		@Path("companyNumber") companyNumber: String,
		@Path("pscId") pscId: String,
	): PersonDto

	@GET(BuildConfig.COMPANIES_HOUSE_GET_PERSONS_CORPORATE_ENDPOINT)
	suspend fun getLegalPerson(
		@Path("companyNumber") companyNumber: String,
		@Path("pscId") pscId: String,
	): PersonDto

}


